src/autogluon/bench/eval/benchmark_context/utils.py

```diff
@@ -1,11 +1,112 @@
+from __future__ import annotations
+
+from urllib.parse import urlparse
+
+import boto3
+
 from autogluon.common.loaders import load_s3
 from autogluon.common.utils import s3_utils
 
 
-def get_s3_paths(path_prefix: str, contains=None, suffix=None):
+def is_s3_url(path: str) -> bool:
+    """
+    Checks if path is a s3 uri.
+
+    Params:
+    -------
+    path: str
+        The path to check.
+
+    Returns:
+    --------
+    bool: whether the path is a s3 uri.
+    """
+    if (path[:2] == "s3") and ("://" in path[:6]):
+        return True
+    return False
+
+
+def get_bucket_key(s3_uri: str) -> tuple[str, str]:
+    """
+    Retrieves the bucket and key from a s3 uri.
+
+    Params:
+    -------
+    origin_path: str
+        The path (s3 uri) to be parsed.
+
+    Returns:
+    --------
+    bucket_name: str
+        the associated bucket name
+    object_key: str
+        the associated key
+    """
+    if not is_s3_url(s3_uri):
+        raise ValueError("Invalid S3 URI scheme. It should be 's3'.")
+
+    parsed_uri = urlparse(s3_uri)
+    bucket_name = parsed_uri.netloc
+    object_key = parsed_uri.path.lstrip("/")
+
+    return bucket_name, object_key
+
+
+def get_s3_paths(path_prefix: str, contains: str | None = None, suffix: str | None = None) -> list[str]:
+    """
+    Gets all s3 paths in the path_prefix that contain 'contains'
+    and end with 'suffix.'
+
+    Params:
+    -------
+    path_prefix: str
+        The path prefix.
+    contains : Optional[str], default = None
+        Can be specified to limit the returned outputs.
+        For example, by specifying the constraint, such as ".1h8c."
+    suffix: str, default = None
+        Can be specified to limit the returned outputs.
+        For example, by specifying "leaderboard.csv" only objects ending
+        with this suffix will be included
+        If no suffix provided, will save all files in artifact directory.
+
+    Returns:
+    --------
+    List[str]: All s3 paths that adhere to the conditions passed in.
+    """
     bucket, prefix = s3_utils.s3_path_to_bucket_prefix(path_prefix)
     objects = load_s3.list_bucket_prefix_suffix_contains_s3(
         bucket=bucket, prefix=prefix, suffix=suffix, contains=contains
     )
     paths_full = [s3_utils.s3_bucket_prefix_to_path(bucket=bucket, prefix=file, version="s3") for file in objects]
     return paths_full
+
+
+def copy_s3_object(origin_path: str, destination_path: str) -> bool:
```

Comment:
QQ: What if the way we store aggregated results changes, by that I mean what if the directory pattern changes - now it is `benchmark_name/module/aggregated/benchmark_name`, lets assume we change this to something else and make it `benchmark_name/module/aggregated/constraint/benchmark_name`?
would it still work?

Reply:
This function would still work fine, since it is only acting on the s3 prefix that has been passed down. However, any changes to where we store the aggregated results would be need to reflected in the `save_path` variable passed into the artifact save methods ([curves](https://github.com/autogluon/autogluon-bench/blob/623c06e7027c9377ea86873161a5c78c8d0189e7/src/autogluon/bench/eval/aggregate/aggregate.py#L53) and [results](https://github.com/autogluon/autogluon-bench/blob/623c06e7027c9377ea86873161a5c78c8d0189e7/src/autogluon/bench/eval/aggregate/aggregate.py#L61)). And this requirement aligns with how the aggregated s3 paths for the results.csv files were previously structured.

Basically the only thing hardcoded is the top level directory order and bottom most directory order in the s3 paths when trying to retrieve objects from the benchmark path.

Regarding the top level directory order, the s3 path where we check for benchmark results must be formatted like this: "s3://{s3_bucket}/{module}/{benchmark_name}", however, this was also the required format prior to my changes. Also note that all of these fields can be specified by the user in the agbench-aggregate-results command.

Regarding the bottom most directory order, when we are given the s3 path to the benchmark_name folder, we recursively search all objects for prefixes ending in `learning_curves.json`. So, this renders all other path structures in between irrelevant. The only aspect that is hardcoded is that once the curve json files are found, it expects them to be stored such that the previous directory is the fold and before that is the dataset name (shown below). This structure is automatically enforced by automlbenchmark when saving artifacts, so regardless of what changes are made here, we will be able to trace backwards to get the fold and dataset for a learning_curves.json file.
```
.
├── dataset_a/
│   ├── 0/
│   │   └── learning_curves.json
│   ├── 1/
│   │   └── learning_curves.json
│   └── 2/
│       └── learning_curves.json
├── dataset_b/
│   ├── 0/
│   │   └── learning_curves.json
│   ├── 1/
│   │   └── learning_curves.json
│   └── 2/
│       └── learning_curves.json
└── ...
```